Add a guide for deploying TAG using Helm

justinas opened the pull request:
Add a guide for deploying the Teleport Access Graph service to Kubernetes clusters using the Helm chart, and integrating the existing Teleport cluster with TAG.

The reference was auto-generated from source code in the `access-graph` repo. There's currently no automation for updating this.

Also bumps the TAG version referenced in the docs to the latest.

Commit:
Add a guide for deploying TAG using Helm

Commit:
Update TAG version to 1.11.0

Commit:
Fix prose lint

justinas commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
Jakub suggested we add an example of generating snakeoil TLS certificates for ease-of-use. I'm not sure if it is a good practice (although we do have a few instances of similar instructions for manually issuing certificates), let me know. For now I kept it with this small disclaimer 🙂

hugoShaka replied:
I'm interested if people have a strong opinion, I have the same problem with the `teleport-cluster` chart for self-hosted examples.

I don't know if `cfssl` is better or worse than raw openssl.

Commit:
Add auto-generated reference for TAG Helm chart

justinas commented on docs/pages/reference/helm-reference/includes/zz_generated.teleport-access-graph.mdx:
This was generated using `render-helm-ref`, source is https://github.com/gravitational/access-graph/blob/main/deploy/helm/teleport-access-graph/values.yaml

ptgott commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
```diff
@@ -0,0 +1,220 @@
+in a Kubernetes cluster using a Helm Chart,
+and enable the Access Graph feature in your Teleport cluster.
+
+The full listing of supported parameters can be found in the [`teleport-access-graph` Helm chart reference](../../reference/helm-reference/teleport-access-graph.mdx)
```
```suggestion
The full listing of supported parameters can be found in the [`teleport-access-graph` Helm chart reference](../../reference/helm-reference/teleport-access-graph.mdx).
```

ptgott commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
```diff
@@ -0,0 +1,220 @@
+The full listing of supported parameters can be found in the [`teleport-access-graph` Helm chart reference](../../reference/helm-reference/teleport-access-graph.mdx)
+
+Teleport Access Graph is enabled by default for Teleport Team users.
+You can skip the setup process here and start using Access Graph right away
```
```suggestion
You can skip the setup process and start using Access Graph right away
```

ptgott commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
```diff
@@ -0,0 +1,220 @@
+    in the same Kubernetes cluster that will be used to deploy Teleport Access Graph .
+- An updated `license.pem` with Teleport Policy enabled.
+- A PostgreSQL database server v14 or later.
+  - Access Graph needs a dedicated [database](https://www.postgresql.org/docs/current/sql-createdatabase.html) to store its data.
```
I would be clearer about what this is linking to: does this mean it requires Postgres? Would it make sense to say we require Postgres without including the link?

justinas replied:
I link to the `CREATE DATABASE` docs because this is my attempt to clarify the overloaded "database" terminology.

In general, "database" can refer to a database server / cluster, i.e. if you `kubectl run postgres`, you have "a postgres database" deployed. But inside a single Postgres server / cluster there can be several "databases", i.e. separate silos for data (the hierarchy being roughly: database -> schema -> table).

In the TAG case, we need a separate database in the Postgres hierarchy sense, i.e. you should not colocate TAG in the same database as your business app (but it *can* be on the same Postgres server / cluster!).

I'd guess that no serious user would try to comingle TAG and a different app in the same Postgres database (in the hierarchy sense), but I tried to clarify this just to be safe 😅

ptgott commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
```diff
@@ -0,0 +1,220 @@
+    The user that TAG connects to the database with needs to be the owner of this database, or have similar broad permissions:
+    at least the `CREATE TABLE` privilege on the `public` schema, and the `CREATE SCHEMA` privilege.
+  - Amazon RDS for PostgreSQL is supported.
+- A TLS certificate for the Access Graph service
```
Might be good to clarify that you can also use a tool like cert-manager to issue certificates automatically (i.e., we don't just require a certificate file).

hugoShaka replied:
I don't think we support cert reloading, so cert-manager will renew certs but TAG will not pick them up.

ptgott commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
I would remove the link here and make it more explicit in the prerequisites that we assume you're using the `teleport-cluster` Helm chart. Right now, we say generally that "we assume that the Teleport cluster is set up using a Helm chart", but we can be more explicit and use the name of the chart there.

ptgott commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
```suggestion
Then, update the values for the deployment of the [teleport-cluster](../reference/helm-reference/teleport-cluster/) Helm chart as follows:
```

ptgott commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
```suggestion
You can find the Access Graph in the "Access Management" tab in the Web UI.
```

hugoShaka commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
````diff
@@ -0,0 +1,220 @@
+echo "Certificates issued successfully"
+```
+
+The script will ask to input the same password 4 times:
+twice to create and confirm the password to encrypt the CA private key material with,
+then once to use the private key to self-sign the CA certificate,
+and finally once to use for signing the TAG server certificate.
````
I'm surprised by the password prompt. I have a very similar openssl script, and it doesn't encrypt the CA. Is that because of the `-aes256` flag? Can we just remove it? I don't think we care about encrypting the CA as this is a snake oil script for local dev anyway.

justinas replied:
I intentionally included the `aes256` flag for the CA cert. This is a snakeoil script, but I still tried to be cautious. We could remove it, it is up to you.

For context, our [HSM guide](https://goteleport.com/docs/choose-an-edition/teleport-enterprise/hsm/#step-15-set-up-your-hsm) includes `aes256`, but it is not positioned as a "snakeoil" script

hugoShaka commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
nit: public is a bit vague. The CA must be in the debian package `ca-certificates`, which is derived from Mozilla's root store.

Also, the CA might also be private but trusted by the auth if they're using the `tls.existingCACertificate` value.

 I think we can drop the whole step 3 cert section and add a requirement:

> Your certificate must be trusted by teleport auth and proxy services. If that's not the case you can set `access_graph.ca: /var/run/access-graph/ca.pem` or if your cluster is deployed by the `teleport-cluster` Helm chart, you can leverage `tls.existingCACertificate`" (the reference contains instructions as how to create the secret).

justinas replied:
I will rephrase to get rid of the "public" wording and shed more light on how Teleport manages trust for CAs.

I'm not 100% sure that it is a good experience for the reader to have to jump to [the reference for `tls.existingCACertificate`](https://goteleport.com/docs/reference/helm-reference/teleport-cluster/#tlsexistingcasecretname).

Correct me if I'm wrong, but it seems that `tls.existingCACertificate` overrides the trusted CA bundle wholesale, i.e. the user might need to make a bundle with their private CAs + the Mozilla bundle or similar? It might be a bit of a footgun with potentially broader impact to the cluster as well if they had not used this option before.

hugoShaka replied:
> Correct me if I'm wrong, but it seems that tls.existingCACertificate overrides the trusted CA bundle wholesale, i.e. the user might need to make a bundle with their private CAs + the Mozilla bundle or similar?

Setting `tls.existingCACertificate` sets `SSL_CERT_FILE`. This only overrides the certs file, not the certs dir. This means that the trusted CAs coming from `/etc/ssl/certs` are still trusted and it won't break your existing trust. This is the recommended way to trust a private CA in the `teleport-cluster` Helm chart.

[Documentation](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_default_verify_paths.html):
> There is one default directory and one default file. The default CA certificates directory is called "certs" in the default OpenSSL directory. Alternatively the SSL_CERT_DIR environment variable can be defined to override this location. The default CA certificates file is called "cert.pem" in the default OpenSSL directory. Alternatively the SSL_CERT_FILE environment variable can be defined to override this location.

justinas replied:
Thanks, I missed the distinction between `SSL_CERT_FILE` and `SSL_CERT_DIR`.

I reworked the section slightly in https://github.com/gravitational/teleport/pull/38266/commits/6231fcce33b9b896b183f7618bfa50c2a453749d

hugoShaka commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
Should be v7

hugoShaka commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
You will want to pin the chart version here, or else people will get TAG upgrades by mistake, which can be incompatible with their Teleport version or bring breaking changes.

hugoShaka commented on docs/pages/reference/helm-reference/teleport-access-graph.mdx:
```diff
@@ -0,0 +1,19 @@
+
+The `teleport-access-graph` Helm chart deploys the Teleport Access Graph service.
+
+See [Run Teleport Access Graph on Self-Hosted Clusters with Helm](../../management/dynamic-resources/teleport-operator-standalone.mdx)
```
bad copy paste, the link must point to the TAG guide

hugoShaka commented on docs/pages/access-controls/access-graph/self-hosted-helm.mdx:
People will copy-paste this without thinking. This includes a hardcoded namespace (which is different from the default one we use in the docs) and a hardcoded release name.

I would remove this code entirely.

Also I'm not sure why you are rolling out auth and proxy deployments. The chart should do it for you on configuration change as the deployment pod spec changes (it contains a config hash). Is there a hidden race condition somewhere requiring to rollout the proxies a second time?

justinas replied:
> This includes a hardcoded namespace (which is different from the default one we use in the docs) and a hardcoded release name.

Interestingly, the [main guide](https://goteleport.com/docs/deploy-a-cluster/helm-deployments/kubernetes-cluster/) uses the `teleport-cluster` namespace, while the [custom config](https://goteleport.com/docs/deploy-a-cluster/helm-deployments/custom/#step-23-setting-up-a-teleport-cluster-with-helm-using-a-custom-config) guide uses `teleport`).

> Also I'm not sure why you are rolling out auth and proxy deployments.
The auth deployment we are not re-rolling, just waiting for the upgrade to complete (which *is* happening due to the config change).

The proxy deployment is due to the fact that proxies only query the TAG feature status (enabled/disabled) from the Auth once, and then caches it indefinitely. The proxy config does not change, but we sadly still need to re-roll it to pick up the change done in Auth 😕 So we still need this step for the user to fully set up and be able to use TAG.

Would it be better if I replaced the hardcoded namespace etc. with `<Var />`-s that people would need to fill in?

hugoShaka replied:
> Interestingly, the [main guide](https://goteleport.com/docs/deploy-a-cluster/helm-deployments/kubernetes-cluster/) uses the teleport-cluster namespace, while the [custom config](https://goteleport.com/docs/deploy-a-cluster/helm-deployments/custom/#step-23-setting-up-a-teleport-cluster-with-helm-using-a-custom-config) guide uses teleport).

Yeah, the custom config is a portage of the legacy guides and uses `teleport` like them. I think I should delete it now that you can do everything with the `teleportConfig` ovveride in the `standalone` mode. Users don't need `scratch`.

> The proxy config does not change, but we sadly still need to re-roll it to pick up the change done in Auth 😕

Ugh, even doing a proxy config change might not 100% reliably enable TAG as the proxy might connect to an old auth. We only protect against this in version upgrade scenarios.

> Would it be better if I replaced the hardcoded namespace etc. with <Var />-s that people would need to fill in?

I suppose that's the best solution, given that we have to restart proxies.

justinas committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

Commit:
Update TAG example role to v7

Commit:
Fix link to TAG Helm guide

Commit:
Pin TAG Helm chart version

Pins to the same version as TAG itself,
Helm chart and TAG are versioned in lockstep

Commit:
Call out the teleport Helm chart more explicitly

Commit:
Bump TAG version to 1.13.0

Expecting it to be released by the time these docs go live.

Commit:
Replace teleport-cluster parameters with Vars

Commit:
Elaborate trust of the TAG CA in Teleport

Commit:
Fix the marker style

Commit:
Fix the link to teleport-cluster reference

justinas committed:
Merge branch 'master' into justinas/tag-helm-docs
